tpm2/structures.go

```diff
@@ -371,3 +371,158 @@ type tpmtSigScheme struct {
 	Scheme Algorithm
 	Hash   Algorithm
 }
+
+// AttestationData contains data attested by TPM commands (like Certify).
+type AttestationData struct {
+	Magic               uint32
+	Type                tpmutil.Tag
+	QualifiedSigner     Name
+	ExtraData           []byte
+	ClockInfo           ClockInfo
+	FirmwareVersion     uint64
+	AttestedCertifyInfo *CertifyInfo
+}
+
+// DecodeAttestationData decode a TPMS_ATTEST message. No error is returned if
+// in has extra trailing data.
+func DecodeAttestationData(in []byte) (*AttestationData, error) {
+	buf := bytes.NewBuffer(in)
+
+	var ad AttestationData
+	if err := tpmutil.UnpackBuf(buf,
+		&ad.Magic,
+		&ad.Type,
+	); err != nil {
+		return nil, fmt.Errorf("decoding Magic/Type: %v", err)
+	}
+	n, err := decodeName(buf)
+	if err != nil {
+		return nil, fmt.Errorf("decoding QualifiedSigner: %v", err)
+	}
+	ad.QualifiedSigner = *n
+	if err := tpmutil.UnpackBuf(buf,
+		&ad.ExtraData,
+		&ad.ClockInfo,
+		&ad.FirmwareVersion,
+	); err != nil {
+		return nil, fmt.Errorf("decoding ExtraData/ClockInfo/FirmwareVersion: %v", err)
+	}
+
+	// Add support for more types as needed.
+	if ad.Type != tagAttestCertify {
+		return nil, fmt.Errorf("only Certify attestation structure is supported, got type 0x%x", ad.Type)
+	}
+	ad.AttestedCertifyInfo, err = decodeCertifyInfo(buf)
+	if err != nil {
+		return nil, fmt.Errorf("decoding AttestedCertifyInfo: %v", err)
+	}
+	return &ad, nil
+}
+
+// CertifyInfo contains Certify-specific data for TPMS_ATTEST.
+type CertifyInfo struct {
+	Name          Name
+	QualifiedName Name
+}
+
+func decodeCertifyInfo(in *bytes.Buffer) (*CertifyInfo, error) {
+	var ci CertifyInfo
+
+	n, err := decodeName(in)
+	if err != nil {
+		return nil, fmt.Errorf("decoding Name: %v", err)
+	}
+	ci.Name = *n
+
+	n, err = decodeName(in)
+	if err != nil {
+		return nil, fmt.Errorf("decoding QualifiedName: %v", err)
+	}
+	ci.QualifiedName = *n
+
+	return &ci, nil
+}
+
+// Name contains a Name for TPM entities. This is either a handle or a digest.
```

Comment:
I don't think the second "Name" should be capitalized here, since it's not referring to the Name struct, but just the common noun "name" with its standard meaning. Also, please clarify for the reader: given the "either", does that mean that this is exclusive: only one of Handle and Digest should be set?

Reply:
Done